### ISSUE TITLE
cross_platform_memmem might have a bug that could result in read beyond packet

### ISSUE THREAD

**Issue by amritanshujohri** [seladb/PcapPlusPlus#542]
There is a crash we are looking at which made me look at the following code, I think there is a problem is here:

```
while (needleLen <= (haystackLen - (ptr - haystack)))
	{
		if (NULL != (ptr = (char*)memchr(ptr, (int)(*needle), haystackLen - (ptr - haystack))))
		{
                        // Fixme:At this point ptr + needlelen might be > haystack + haystackLen ), and we might end up reading 
                        // beyond valid datalength
			if (0 == memcmp(ptr, needle, needleLen))
				return ptr;
			else
				++ptr;
		}
		else
			break;
	}
```

Let me know if this is making sense?

**Comment by seladb** [comment]
Yes, I think you're correct. Would you consider making the fix and create a PR?

We can either fix this code or use one of the open-source implementations of this method out there.

**Comment by amritanshujohri** [comment]
The fix at my end is work in progress!

**Comment by seladb** [comment]
@amritanshujohri did you make any progress on that?

**Comment by seladb** [comment]
Fixed in PR #551 . Thank you so much @amritanshujohri for your contribution!
